watch-purge command to stop watching all roots at once

phmx opened the pull request:
watch-purge command is like watch-del but removes all watched roots

wez commented on cmds/watch.c:
```diff
@@ -97,6 +97,37 @@ static void cmd_watch_list(struct watchman_client *client, json_t *args)
+  set_prop(resp, "purged", purged);
+  send_and_dispose_response(client, resp);
+}
+W_CMD_REG("watch-purge", cmd_watch_purge, CMD_DAEMON, NULL)
```
can we call it `watch-del-all`?  Purge has a slightly different connotation

wez commented on cmds/watch.c:
```diff
@@ -97,6 +97,37 @@ static void cmd_watch_list(struct watchman_client *client, json_t *args)
+  size_t  root_paths_size, i;
+  unused_parameter(args);
+
+  root_paths = w_root_watch_list_to_json();
```
Could you add a `w_root_stop_all()` function to root.c that combines the iteration and the stop requests under the root_lock mutex?  That way this command becomes more atomic.  The `w_root_stop_all` function could return a json_t array listing the names that it stopped

wez commented:
Would be good to get this covered by integration tests too.

The `WatchmanTestCase::didRunTests` function cleans up the watches established during a test run; I think we could have that run `watch-del-all` instead.
https://facebook.github.io/watchman/contributing.html has more information about running the integration tests.

Thanks!

stefanpenner commented:
yay

phmx commented:
> can we call it watch-del-all? Purge has a slightly different connotation

Sure, done.

> Could you add a w_root_stop_all() function to root.c that combines the iteration and the stop requests under the root_lock mutex?

Ok, added `json *w_root_stop_watch_all(void)' function to root.c. It returns a json array of previously watched roots.

> Would be good to get this covered by integration tests too.

I'll look into it, thanks for the tip.

> yay, now i can retire: (for root in $(watchman watch-list | jq '.roots | .[]'); do watchman watch-del $(echo $root | sed 's/"//g') ; done)

I can see Your point, maybe command atomicity will add some value for You. Also, I think that simple `watch-del-all' command is more convenient than some shell magic, though less Unixy.

wez commented on arcanist/lib/WatchmanTestCase.php:
```diff
@@ -139,13 +139,12 @@ function willRunTests() {
-        // Swallow
-      }
+    try {
+      $this->watchmanCommand('watch-del-all');
```
it would be great to have a more proactive/positive assertion that this is functioning; could you add a new test in `tests/integration` that watches a couple of dirs then asserts that all of those were deleted?

wez commented on root.c:
```diff
@@ -2083,6 +2083,32 @@ bool w_root_stop_watch(w_root_t *root)
+    w_string_t *path = root->root_path;
+    json_t *json_root = json_string_binary(path->buf, path->len);
+
+    if (w_ht_iter_del(watched_roots, &iter)) {
```
there's a potential race here; root may have its final ref released by the delete operation, making it unsafe to call w_root_cancel on it below.  You'll need to `w_root_addref` prior to calling this, and then `w_root_delref` before the next iteration

phmx commented:
> You'll need to w_root_addref prior to calling this, and then w_root_delref before the next iteration

Fixed, thanks for review.

> could you add a new test in tests/integration that watches a couple of dirs then asserts that all of those were deleted?

Done. Added `tests/integration/watch-del-all.php`. It actually has detected some bugs in `w_root_stop_watch_all` implementation calling `w_ht_iter_del` (looks like it didn't remove some roots when called concurrently), so I've used `w_ht_del` once again.

wez commented on root.c:
```diff
@@ -2083,6 +2083,41 @@ bool w_root_stop_watch(w_root_t *root)
+    w_root_delref(root);
+  }
+  pthread_mutex_unlock(&root_lock);
+
```
need to `free(roots);` here

wez commented on tests/integration/watch-del-all.php:
```diff
@@ -0,0 +1,44 @@
+
+    $this->assertEqual($dirs, $watched);
+
+    $this->watchmanCommand('watch-del-all');
```
I think you can make an assertion here about the return value of this; it should match `$watched`

wez commented on tests/integration/watch-del-all.php:
```diff
@@ -0,0 +1,44 @@
+
+    $this->watchmanCommand('watch-del-all');
+
+    $this->waitForWatchman(
```
I'd expect this to be an empty list immediately because the implementation is removing from the watch list immediately; I think you can implement this as a simple assertEquals on the result `watchmanCommand`?

phmx commented:
> need to free(roots); here

Fixed.

> I think you can make an assertion here about the return value of this; it should match $watched

Yes, done.

> I think you can implement this as a simple assertEquals on the result watchmanCommand?

 Fixed. It was a legacy of struggle with `w_ht_iter_del` =)

wez commented on root.c:
```diff
@@ -2083,6 +2083,42 @@ bool w_root_stop_watch(w_root_t *root)
   return stopped;
 }
 
+json_t *w_root_stop_watch_all(void)
```
```
/Users/wez/fb/watchman/.git/rebase-apply/patch:51: trailing whitespace.
json_t *w_root_stop_watch_all(void)
warning: 1 line adds whitespace errors.
```

wez commented on tests/integration/watch-del-all.php:
```diff
@@ -0,0 +1,45 @@
+      }
+
+      $this->watch($d);
+      $this->assertFileList($d, $files);
```
```
/Users/wez/fb/watchman/.git/rebase-apply/patch:37: trailing whitespace.
      $this->assertFileList($d, $files);
```

wez commented on tests/integration/watch-del-all.php:
```diff
@@ -0,0 +1,45 @@
+    $dirs = array_map(
+      function ($file) use($root) {
+        return "$root/$file";
+      },
```
```
/Users/wez/fb/watchman/.git/rebase-apply/patch:25: trailing whitespace.
      },
```

wez commented:
This is looking really good, thanks!
Can you fix up the trailing whitespace nits (see inline comments), and one last thing to add is documentation for the command.
You can add `docs/cmd.watch-del-all.markdown` with appropriate content (model it after `docs/cmd.watch-del.markdown`).
You should be able to preview your docs changes locally by doing:

```
cd website
npm start
```

and it will print out a URL that you can open in your browser.

Thanks!

phmx commented:
> Can you fix up the trailing whitespace nits (see inline comments), ...

Hmm, it should have been fixed by commits https://github.com/phmx/watchman/commit/07db0a81e494f479c42c70470d95a23fb0158329 and https://github.com/phmx/watchman/commit/a9429f223b6dba163c9bfa975a4911d51f11b0be. I don't see any trailing whitespaces at the tip of the branch.

> and one last thing to add is documentation for the command.

Done. I added `docs/cmd.watch-del-all.markdown`, but didn't find a `website` directory You mentioned. Is it among projects on https://github.com/facebook page? Searched there too to no avail.

wez commented:
re: whitespace, could just be me misreading the output; I use `hub` to apply the pull request and I saw those warnings.

re: website, I'd forgotten that that stuff isn't in this repo.

Sorry for the churn!  Just go ahead and add the markdown file and I'll take it from here, thanks!

phmx commented:
My apologies for the delay.

> Sorry for the churn! Just go ahead and add the markdown file and I'll take it from here, thanks!

I've done it it with https://github.com/phmx/watchman/commit/1a3e5d2b29026cbbd838b6e6190b0649cff79c4b commit. Is there something wrong?

wez commented:
Thanks!
I've squashed the commits and pushed this as 7ffa6310feabed987525cb1d5946cdcc9169d79b.
I also bumped the version number up to 3.1.1 so that we can indicate availability correctly in the docs.
